### ISSUE TITLE
jetty:9-jre8 polluted with class version 53.0 (jre 9)

### ISSUE THREAD

**Issue by draeath** [eclipse/jetty.docker#49]
It seems that `jetty:9-jre8` (`5460fbba9ad6`) is broken and was built with JDK9 as a minimum runtime JRE requirement:

```
2021-05-04 20:57:42.000:INFO:docker-entrypoint:jetty start from /var/lib/jetty/jetty.start
2021-05-04 20:57:43.304:INFO::main: Logging initialized @553ms to org.eclipse.jetty.util.log.StdErrLog
2021-05-04 20:57:44.083:WARN:oejx.XmlConfiguration:main: Config error at <Call name="addConnectionFactory"><Arg>|      <New id="alpn" class="org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory"><Arg name="protocols" type="String"><Property name="jetty.alpn.protocols" deprecated="alpn.protocols" default=""/></Arg><Set name="defaultProtocol"><Property name="jetty.alpn.defaultProtocol" deprecated="alpn.defaultProtocol"/></Set></New>|    </Arg></Call> java.lang.reflect.InvocationTargetException in file:///usr/local/jetty/etc/jetty-alpn.xml
2021-05-04 20:57:44.084:WARN:oejx.XmlConfiguration:main: 
java.security.PrivilegedActionException: java.lang.reflect.InvocationTargetException
        at java.security.AccessController.doPrivileged(Native Method)
        at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1857)
Caused by: 
java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.invokeConstructor(XmlConfiguration.java:752)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.construct(XmlConfiguration.java:1039)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.newObj(XmlConfiguration.java:1011)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.itemValue(XmlConfiguration.java:1540)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.value(XmlConfiguration.java:1441)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.access$700(XmlConfiguration.java:395)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration$Args.<init>(XmlConfiguration.java:1699)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration$Args.<init>(XmlConfiguration.java:1686)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:942)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:515)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:468)
        at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:380)
        at org.eclipse.jetty.xml.XmlConfiguration.lambda$main$3(XmlConfiguration.java:1893)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1857)
Caused by: 
java.lang.IllegalStateException: No Server ALPNProcessors!
        at org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory.<init>(ALPNServerConnectionFactory.java:52)
        at org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory.<init>(ALPNServerConnectionFactory.java:45)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.invokeConstructor(XmlConfiguration.java:752)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.construct(XmlConfiguration.java:1039)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.newObj(XmlConfiguration.java:1011)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.itemValue(XmlConfiguration.java:1540)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.value(XmlConfiguration.java:1441)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.access$700(XmlConfiguration.java:395)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration$Args.<init>(XmlConfiguration.java:1699)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration$Args.<init>(XmlConfiguration.java:1686)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:942)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:515)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:468)
        at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:380)
        at org.eclipse.jetty.xml.XmlConfiguration.lambda$main$3(XmlConfiguration.java:1893)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1857)
Suppressed: 
        |java.lang.UnsupportedClassVersionError: org/eclipse/jetty/alpn/java/server/JDK9ServerALPNProcessor has been compiled by a more recent version of the Java Runtime (class file version 53.0), this version of the Java Runtime only recognizes class file versions up to 52.0
        |       at java.lang.ClassLoader.defineClass1(Native Method)
        |       at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
        |       at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
        |       at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
        |       at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
        |       at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
        |       at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
        |       at java.security.AccessController.doPrivileged(Native Method)
        |       at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
        |       at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
        |       at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
        |       at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
        |       at java.lang.Class.forName0(Native Method)
        |       at java.lang.Class.forName(Class.java:348)
        |       at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:370)
        |       at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
        |       at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
        |       at org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory.<init>(ALPNServerConnectionFactory.java:59)
        |       at org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory.<init>(ALPNServerConnectionFactory.java:45)
        |       at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        |       at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        |       at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        |       at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.invokeConstructor(XmlConfiguration.java:752)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.construct(XmlConfiguration.java:1039)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.newObj(XmlConfiguration.java:1011)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.itemValue(XmlConfiguration.java:1540)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.value(XmlConfiguration.java:1441)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.access$700(XmlConfiguration.java:395)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration$Args.<init>(XmlConfiguration.java:1699)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration$Args.<init>(XmlConfiguration.java:1686)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:942)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:515)
        |       at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:468)
        |       at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:380)
        |       at org.eclipse.jetty.xml.XmlConfiguration.lambda$main$3(XmlConfiguration.java:1893)
        |       at java.security.AccessController.doPrivileged(Native Method)
        |       at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1857)
Exception in thread "main" java.security.PrivilegedActionException: java.lang.reflect.InvocationTargetException
        at java.security.AccessController.doPrivileged(Native Method)
        at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1857)
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.invokeConstructor(XmlConfiguration.java:752)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.construct(XmlConfiguration.java:1039)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.newObj(XmlConfiguration.java:1011)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.itemValue(XmlConfiguration.java:1540)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.value(XmlConfiguration.java:1441)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.access$700(XmlConfiguration.java:395)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration$Args.<init>(XmlConfiguration.java:1699)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration$Args.<init>(XmlConfiguration.java:1686)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:942)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:515)
        at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:468)
        at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:380)
        at org.eclipse.jetty.xml.XmlConfiguration.lambda$main$3(XmlConfiguration.java:1893)
        ... 2 more
Caused by: java.lang.IllegalStateException: No Server ALPNProcessors!
        at org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory.<init>(ALPNServerConnectionFactory.java:52)
        at org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory.<init>(ALPNServerConnectionFactory.java:45)
        ... 19 more
        Suppressed: java.lang.UnsupportedClassVersionError: org/eclipse/jetty/alpn/java/server/JDK9ServerALPNProcessor has been compiled by a more recent version of the Java Runtime (class file version 53.0), this version of the Java Runtime only recognizes class file versions up to 52.0
                at java.lang.ClassLoader.defineClass1(Native Method)
                at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
                at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
                at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
                at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
                at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
                at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
                at java.security.AccessController.doPrivileged(Native Method)
                at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
                at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
                at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
                at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
                at java.lang.Class.forName0(Native Method)
                at java.lang.Class.forName(Class.java:348)
                at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:370)
                at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
                at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
                at org.eclipse.jetty.alpn.server.ALPNServerConnectionFactory.<init>(ALPNServerConnectionFactory.java:59)
                ... 20 more
```

My usage:

```Dockerfile
FROM jetty:9-jre8 AS runtime

USER root
RUN mkdir -pv /mnt/jetty-logs && chown -v jetty:jetty /mnt/jetty-logs

USER jetty
ENV JAVA_OPTIONS "-Xmx512m -Xms512m"
RUN java -jar ${JETTY_HOME}/start.jar --add-to-start=https,http2 --approve-all-licenses
RUN rm -v ${JETTY_BASE}/start.d/http.ini
ADD --chown=jetty:jetty cas-redirect.war /var/lib/jetty/webapps/ROOT.war
COPY --chown=jetty:jetty --from=build /root/workdir/cas.war /var/lib/jetty/webapps/cas.war
ADD --chown=jetty:jetty etc/cas /etc/cas
ADD --chown=jetty:jetty keystore /var/lib/jetty/etc/keystore
ADD --chown=jetty:jetty ssl.ini /var/lib/jetty/start.d/ssl.ini
```

The referenced .war files were built in a `openjdk:8-jre` container.

**Comment by joakime** [comment]
I can replicate easily.

Short form of Dockerfile to replicate ...

``` Dockerfile
FROM jetty:9-jre8 AS runtime

USER root

USER jetty
ENV JAVA_OPTIONS "-Xmx512m -Xms512m"
RUN java -jar ${JETTY_HOME}/start.jar --add-to-start=https,http2, --approve-all-licenses
RUN rm -v ${JETTY_BASE}/start.d/http.ini
```

There is something else going on.
The suppressed exception you are getting is expected and normal when running on JDK 8.

The Jetty 9.4.x release is built with OpenJDK 11 and is source/target of 1.8 for most modules, but not 100% of them.
There are some modules with source/target of JDK 9.
The `org/eclipse/jetty/alpn/java/server/JDK9ServerALPNProcessor` is one of them, and is included in the Jetty home tarball intentionally.

When Jetty's HTTP2 module loads up the [`ALPNServerConnectionFactory` will use the Java `ServiceLoader` to load all `org.eclipse.jetty.io.ssl.ALPNProcessor.Server` found on the classloader](https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-alpn/jetty-alpn-server/src/main/java/org/eclipse/jetty/alpn/server/ALPNServerConnectionFactory.java#L49-L86)

This includes Conscrypt, JDK9, OpenJDK8, etc ...

If the defined `org.eclipse.jetty.io.ssl.ALPNProcessor.Server` cannot load, it's ignored and logged as suppressed.
If the call to `org.eclipse.jetty.io.ssl.ALPNProcessor.Server.init()` fails, it's ignored and logged as suppressed.

If there are no `ALPNProcessor.Server` implementations left, the exception you see is produced.
The reported exception shows that only 1 `ALPNProcessor.Server` was loaded by the ServiceLoader and was attempted to be located, and that's the problem that needs to be resolved, where are the other ones?

The resulting config looks like this ...

``` none
$ docker run --rm -it jetty:issue-49 bash
jetty@3658ca792246:~$ java --version
Unrecognized option: --version
Error: Could not create the Java Virtual Machine.
Error: A fatal exception has occurred. Program will exit.
jetty@3658ca792246:~$ java -version
openjdk version "1.8.0_292"
OpenJDK Runtime Environment (build 1.8.0_292-b10)
OpenJDK 64-Bit Server VM (build 25.292-b10, mixed mode)
jetty@3658ca792246:~$ cd ${JETTY_BASE}
jetty@3658ca792246:~$ pwd
/var/lib/jetty
jetty@3658ca792246:~$ java -jar $JETTY_HOME/start.jar --list-config

Java Environment:
-----------------
 java.home = /usr/local/openjdk-8 (null)
 java.vm.vendor = Oracle Corporation (null)
 java.vm.version = 25.292-b10 (null)
 java.vm.name = OpenJDK 64-Bit Server VM (null)
 java.vm.info = mixed mode (null)
 java.runtime.name = OpenJDK Runtime Environment (null)
 java.runtime.version = 1.8.0_292-b10 (null)
 java.io.tmpdir = /tmp (null)
 user.dir = /var/lib/jetty (null)
 user.language = en (null)
 user.country = null (null)

Jetty Environment:
-----------------
 jetty.version = 9.4.40.v20210413
 jetty.tag.version = jetty-9.4.40.v20210413
 jetty.build = b881a572662e1943a14ae12e7e1207989f218b74
 jetty.home = /usr/local/jetty
 jetty.base = /var/lib/jetty

Config Search Order:
--------------------
 <command-line>
 ${jetty.base} -> /var/lib/jetty
 ${jetty.home} -> /usr/local/jetty


JVM Arguments:
--------------
 (no jvm args specified)

System Properties:
------------------
 (no system properties specified)

Properties:
-----------
 java.version = 1.8.0_292
 java.version.major = 1
 java.version.micro = 0
 java.version.minor = 8
 java.version.platform = 8
 jetty.base = /var/lib/jetty
 jetty.base.uri = file:///var/lib/jetty
 jetty.home = /usr/local/jetty
 jetty.home.uri = file:///usr/local/jetty
 runtime.feature.alpn = true

Jetty Server Classpath:
-----------------------
Version Information on 42 entries in the classpath.
Note: order presented here is how they would appear on the classpath.
      changes to the --module=name command line options will be reflected here.
 0:                    (dir) | ${jetty.base}/resources
 1:                    3.1.0 | ${jetty.home}/lib/servlet-api-3.1.jar
 2:                 3.1.0.M0 | ${jetty.home}/lib/jetty-schemas-3.1.jar
 3:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-http-9.4.40.v20210413.jar
 4:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-server-9.4.40.v20210413.jar
 5:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-xml-9.4.40.v20210413.jar
 6:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-util-9.4.40.v20210413.jar
 7:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-io-9.4.40.v20210413.jar
 8:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-alpn-java-server-9.4.40.v20210413.jar
 9:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-alpn-server-9.4.40.v20210413.jar
10:      1.4.1.v201005082020 | ${jetty.home}/lib/mail/javax.mail.glassfish-1.4.1.v201005082020.jar
11:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-jndi-9.4.40.v20210413.jar
12:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-security-9.4.40.v20210413.jar
13:                      1.3 | ${jetty.home}/lib/transactions/javax.transaction-api-1.3.jar
14:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-servlet-9.4.40.v20210413.jar
15:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-webapp-9.4.40.v20210413.jar
16:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-plus-9.4.40.v20210413.jar
17:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-annotations-9.4.40.v20210413.jar
18:                      9.0 | ${jetty.home}/lib/annotations/asm-9.0.jar
19:                      9.0 | ${jetty.home}/lib/annotations/asm-analysis-9.0.jar
20:                      9.0 | ${jetty.home}/lib/annotations/asm-commons-9.0.jar
21:                      9.0 | ${jetty.home}/lib/annotations/asm-tree-9.0.jar
22:                    1.3.2 | ${jetty.home}/lib/annotations/javax.annotation-api-1.3.2.jar
23:    3.19.0.v20190903-0936 | ${jetty.home}/lib/apache-jsp/org.eclipse.jdt.ecj-3.19.0.jar
24:         9.4.40.v20210413 | ${jetty.home}/lib/apache-jsp/org.eclipse.jetty.apache-jsp-9.4.40.v20210413.jar
25:                   8.5.54 | ${jetty.home}/lib/apache-jsp/org.mortbay.jasper.apache-el-8.5.54.jar
26:                   8.5.54 | ${jetty.home}/lib/apache-jsp/org.mortbay.jasper.apache-jsp-8.5.54.jar
27:                    1.2.5 | ${jetty.home}/lib/apache-jstl/org.apache.taglibs.taglibs-standard-impl-1.2.5.jar
28:                    1.2.5 | ${jetty.home}/lib/apache-jstl/org.apache.taglibs.taglibs-standard-spec-1.2.5.jar
29:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-client-9.4.40.v20210413.jar
30:         9.4.40.v20210413 | ${jetty.home}/lib/jetty-deploy-9.4.40.v20210413.jar
31:         9.4.40.v20210413 | ${jetty.home}/lib/http2/http2-common-9.4.40.v20210413.jar
32:         9.4.40.v20210413 | ${jetty.home}/lib/http2/http2-hpack-9.4.40.v20210413.jar
33:         9.4.40.v20210413 | ${jetty.home}/lib/http2/http2-server-9.4.40.v20210413.jar
34:                      1.0 | ${jetty.home}/lib/websocket/javax.websocket-api-1.0.jar
35:         9.4.40.v20210413 | ${jetty.home}/lib/websocket/javax-websocket-client-impl-9.4.40.v20210413.jar
36:         9.4.40.v20210413 | ${jetty.home}/lib/websocket/javax-websocket-server-impl-9.4.40.v20210413.jar
37:         9.4.40.v20210413 | ${jetty.home}/lib/websocket/websocket-api-9.4.40.v20210413.jar
38:         9.4.40.v20210413 | ${jetty.home}/lib/websocket/websocket-client-9.4.40.v20210413.jar
39:         9.4.40.v20210413 | ${jetty.home}/lib/websocket/websocket-common-9.4.40.v20210413.jar
40:         9.4.40.v20210413 | ${jetty.home}/lib/websocket/websocket-server-9.4.40.v20210413.jar
41:         9.4.40.v20210413 | ${jetty.home}/lib/websocket/websocket-servlet-9.4.40.v20210413.jar

Jetty Active XMLs:
------------------
 ${jetty.home}/etc/jetty-bytebufferpool.xml
 ${jetty.home}/etc/jetty-threadpool.xml
 ${jetty.home}/etc/jetty.xml
 ${jetty.home}/etc/jetty-ssl.xml
 ${jetty.home}/etc/jetty-ssl-context.xml
 ${jetty.home}/etc/jetty-alpn.xml
 ${jetty.home}/etc/jetty-webapp.xml
 ${jetty.home}/etc/jetty-plus.xml
 ${jetty.home}/etc/jetty-annotations.xml
 ${jetty.home}/etc/jetty-deploy.xml
 ${jetty.home}/etc/jetty-http2.xml
 ${jetty.home}/etc/jetty-https.xml
```

**Comment by joakime** [comment]
The wrong alpn implementation was chosen, somehow, huh?

``` shell
$ java -jar $JETTY_HOME/start.jar --lib=lib/jetty-alpn-openjdk8-server-$JETTY_VERSION.jar
2021-05-04 22:38:36.258:INFO::main: Logging initialized @377ms to org.eclipse.jetty.util.log.StdErrLog
2021-05-04 22:38:36.516:INFO:oejs.Server:main: jetty-9.4.40.v20210413; built: 2021-04-13T20:42:42.668Z; git: b881a572662e1943a14ae12e7e1207989f218b74; jvm 1.8.0_292-b10
2021-05-04 22:38:36.531:INFO:oejdp.ScanningAppProvider:main: Deployment monitor [file:///var/lib/jetty/webapps/] at interval 1
2021-05-04 22:38:36.568:INFO:oejus.SslContextFactory:main: x509=X509@eafc191(jetty,h=[jetty.eclipse.org],w=[]) for Server@612679d6[provider=null,keyStore=file:///var/lib/jetty/etc/keystore,trustStore=file:///var/lib/jetty/etc/keystore]
2021-05-04 22:38:36.569:INFO:oejus.SslContextFactory:main: x509=X509@11758f2a(mykey,h=[],w=[]) for Server@612679d6[provider=null,keyStore=file:///var/lib/jetty/etc/keystore,trustStore=file:///var/lib/jetty/etc/keystore]
2021-05-04 22:38:36.640:INFO:oejs.AbstractConnector:main: Started ServerConnector@450090b9{SSL, (ssl, alpn, h2, http/1.1)}{0.0.0.0:8443}
2021-05-04 22:38:36.642:INFO:oejs.Server:main: Started @762ms
```

**Comment by joakime** [comment]
@sbordet any thoughts?

**Comment by joakime** [comment]
I think we should not be using Oracle 8u292, and should be using OpenJDK from AdoptOpenJDK 8u292 instead.

I don't think the ALPN layer was properly backported into the Oracle releases, at least the behavior makes it seem like that is true.

**Comment by sbordet** [comment]
@joakime @lachlan-roberts if you are targeting an image for Java 8 _only_, then you should exclude the dependency on `jetty-alpn-java-server` and add instead the dependency to `jetty-alpn-openjdk8-server`.

If you are targeting an image for Java 8+ (meaning running it with Java 8 or any later major version), then you should have both as dependencies.

If you are targeting an image for Java 9+, then you should only have the dependency on `jetty-alpn-java-server`.

**Comment by draeath** [comment]
Hey folks.

I've run into this again. I have an old build of some garbage I have to keep on life-support and it has a hard requirement on JRE8.

I rebuilt a container today to fix the log4shell CVE and `jetty:9-jre8` is again/still bailing out with:

    java.lang.UnsupportedClassVersionError: org/eclipse/jetty/alpn/java/server/JDK9ServerALPNProcessor has been compiled by a more recent version of the Java Runtime (class file version 53.0), this version of the Java Runtime only recognizes class file versions up to 52.0

**Comment by lachlan-roberts** [comment]
@draeath I think if you add the jetty-alpn-openjdk8-server dependency it should work. Try this:

```Dockerfile
RUN java -jar ${JETTY_HOME}/start.jar --add-to-start=ext && \
         cp ${JETTY_HOME}/lib/jetty-alpn-openjdk8-server-*.jar ${JETTY_BASE}/lib/ext/

```

**Comment by draeath** [comment]
@lachlan-roberts you are a hero. This did the trick, thank you!

I'll leave this open for now, but if y'all think that's sufficient to close this (ie, should I have known to do this as an end user) we can certainly do this.

**Comment by lachlan-roberts** [comment]
Now that the fix from https://github.com/eclipse/jetty.project/issues/7266 is merged and available since `9.4.45`, you should not need the workaround of manually adding the `jetty-alpn-openjdk8-server` jars.
